File: session-01/task-04/word-index.js
const LINES_PRE_PAGE = 45;
const STOP_FREQUENCY_LIMIT = 100;

const fs = require('fs');

const filePath = process.argv[2];

function lineIndexToPage(lineIndex) {
    return Math.floor(lineIndex / LINES_PRE_PAGE) + 1;
}

try {
    const data = fs.readFileSync(filePath, 'utf8');

    const lines = data.split('\n');

    const wordsToOccurrences = new Map();
    for (let i = 0; i < lines.length; i++) {
        const page = lineIndexToPage(i);

        const words = lines[i].split(' ');

        for (const word of words) {
            if (word) {
                const occurrences = wordsToOccurrences.get(word) || [];

                if (!occurrences.includes(page)) {
                    occurrences.push(page);
                }

                wordsToOccurrences.set(word, occurrences);
            }
        }
    }

    const sortedKeys = Array.from(wordsToOccurrences.keys()).sort();

    for (const key of sortedKeys) {
        const occurrences = wordsToOccurrences.get(key);

        if (occurrences.length <= STOP_FREQUENCY_LIMIT) {
            console.log(`${key} ${occurrences.join(', ')}`);
        }
    }

} catch (err) {
    console.error(err);
}

export {lineIndexToPage}

File: session-01/task-02/word_index.py
#!/usr/bin/env python

# Program Description:
#
# Word Index is a program that takes a plain text file as input and
# outputs all the words contained in it
# sorted alphabetically along with the page numbers on which they occur.
# The program assumes that a page is a
# sequence of 45 lines, each line has max 80 characters, and there is no
# hyphenation. Additionally, Word Index
# must ignore all words that occur more than 100 times.

####################################
# Free Flow, a.k.a. No Style
####################################

import sys, string

# Global "constants"
LINES_PER_PAGE = 45
MAX_SIZE_LINE = 80
STOP_FREQUENCY_LIMIT = 100


def lineIndexToPage(lineIndex):
    return lineIndex / LINES_PER_PAGE + 1


# Defining a main method makes testing easier
def main(file_path):
    file = open(file_path)

    wordsToOccurrences = {}
    lineIndex = 0
    for line in file:
        line = line.replace('\n', '')
        page = lineIndexToPage(lineIndex)
        words = line.split(' ')

        for word in words:
            if word:
                if word in wordsToOccurrences:
                    occurrences = wordsToOccurrences[word]
                else:
                    occurrences = []

                if page not in occurrences:
                    occurrences.append(page)

                wordsToOccurrences[word] = occurrences

    words = sorted(wordsToOccurrences.keys())

    for word in words:
        occurrences = wordsToOccurrences[word]

        if len(occurrences) <= STOP_FREQUENCY_LIMIT:
            print(word, wordsToOccurrences[word])



if __name__ == "__main__":
    main(sys.argv[1])
